api/src/main/java/io/cloudevents/message/impl/BaseGenericBinaryMessageImpl.java

```diff
@@ -0,0 +1,71 @@
+package io.cloudevents.message.impl;
+
+import io.cloudevents.SpecVersion;
+import io.cloudevents.message.BinaryMessageVisitor;
+import io.cloudevents.message.BinaryMessageVisitorFactory;
+import io.cloudevents.message.MessageVisitException;
+
+import java.util.Objects;
+import java.util.function.BiConsumer;
+
+/**
+ * This class implements a BinaryMessage, providing commong logic to most protocol bindings
+ * which supports both Binary and Structured mode.
+ * Content-type is handled separately using a key not prefixed with CloudEvents header prefix.
+ *
+ * @param <HV> Header value type
+ */
+public abstract class BaseGenericBinaryMessageImpl<HV> extends BaseBinaryMessage {
+
+    private final SpecVersion version;
+    private final byte[] body;
+
+    protected BaseGenericBinaryMessageImpl(SpecVersion version, byte[] body) {
+        Objects.requireNonNull(version);
+        this.version = version;
+        this.body = body;
+    }
+
+    @Override
+    public <T extends BinaryMessageVisitor<V>, V> V visit(BinaryMessageVisitorFactory<T, V> visitorFactory) throws MessageVisitException, IllegalStateException {
+        BinaryMessageVisitor<V> visitor = visitorFactory.createBinaryMessageVisitor(this.version);
+
+        // Grab from headers the attributes and extensions
+        this.forEachHeader((k, v) -> {
```

Comment:
can it be `key` and `value`?

Reply:
@slinkydeveloper I will appreciate if you let me resolve the conversations I started. Not really sure why this one is marked as resolved